cmd/snap-bootstrap/partition/disks.go

```diff
@@ -0,0 +1,499 @@
+
+	pattern := fmt.Sprintf("%s/dev/block/%d:*", dirs.SysfsDir, d.major)
+	allDevices, err := filepath.Glob(pattern)
+	if err != nil {
+		return nil, fmt.Errorf("internal error: %v", err)
+	}
+
+	// glob does not guarantee a sort, but we need our list of devices sorted
+	sort.Strings(allDevices)
+
+	// populate all the partitions from our candidate devices
+	for _, dev := range allDevices {
+		base := filepath.Base(dev)
+		maj, min, err := parseDeviceMajorMinor(base)
+		if err != nil {
+			continue
+		}
+
+		// ignore any devices that have minor numbers less than the disk itself
+		// the disk will have the same minor so ignore that too
+		if min <= d.minor {
+			continue
+		}
+
+		// now if there is a partition file, this is a partition for our disk
+		if osutil.FileExists(filepath.Join(dev, "partition")) {
+			// read the uevent file for this partition to get the devname
+			p := &partition{
+				major: maj,
+				minor: min,
+			}
+			f, err := os.Open(filepath.Join(dev, "uevent"))
+			if err != nil {
+				continue
+			}
+
+			// now get the full set of udev properties for this partition
+			eventProps, err := parseUdevProperties(f)
+			if err != nil {
+				continue
+			}
+
+			// get the name of the device path to call udevadm
+			// note that here DEVNAME is coming from sysfs's uevent, and for
+			// whatever reason it is just the basename a la vda3, but DEVNAME
+			// when  requested from udevadm proper is the full path a la
+			// /dev/vda3
+			if name, ok := eventProps["DEVNAME"]; ok {
+				p.path = filepath.Join("/dev", name)
+			} else {
+				continue
+			}
+
+			props, err := udevProperties(p.path)
+			if err != nil && props == nil {
+				// only error here if we didn't get a map, we validate the map
+				// in the next steps
+				continue
+			}
+
+			// get the label
+			if labelEncoded, ok := props["ID_FS_LABEL_ENC"]; ok {
+				p.label = labelEncoded
+			} else {
+				continue
+			}
+
+			// finally get the partition uuid
+			if partuuid, ok := props["ID_PART_ENTRY_UUID"]; ok {
+				p.partuuid = partuuid
+			} else {
+				continue
+			}
+
+			d.partitions = append(d.partitions, p)
+		} else {
+			// if there was not a partition file, we hit another disk and must
+			// stop searching (the disk we are looking at will be ignored with
+			// the minor number <= check above)
+			break
+		}
+	}
+
+	// if we didn't find any partitions from above then return an error
+	if len(d.partitions) == 0 {
+		return nil, fmt.Errorf("no partitions found for disk %s", majorMinor)
+	}
+
+	return d, nil
+}
+
+func (d *disk) FindMatchingPartitionUUID(label string) (string, error) {
+	// iterate over the partitions looking for the specified label
+	for _, part := range d.partitions {
+		if part.label == label {
+			return part.partuuid, nil
+		}
+	}
+
+	return "", fmt.Errorf("couldn't find label %q", label)
+}
+
+func (d *disk) MountPointIsFromDisk(mountpoint string, opts *Options) (bool, error) {
+	d2, err := diskFromMountPointImpl(mountpoint, opts)
```

Comment:
if d2 is d this will parse all the partitions again for no good reason, we probably want to split finding the major:minor of a mountpoint vs enumerating the partitions